### ISSUE TITLE
Question about the double added gradients of TV loss.

### ISSUE THREAD

**Issue by xfanhe** [wutong16/Voxurf#42]
Hi, thanks for your great project.
I just find something strange. As below, the tv loss is calculated without cuda at first, and then the gradients of it are added inplace, according to the DVGOv2.
```
        if global_step>cfg_train.tv_from and global_step<cfg_train.tv_end and global_step%cfg_train.tv_every==0:
            if cfg_train.weight_tv_density>0:
                tv_terms = getattr(cfg_train, 'tv_terms', dict())
                sdf_tv, smooth_grad_tv = tv_terms['sdf_tv'], tv_terms['smooth_grad_tv']
                if smooth_grad_tv > 0:
                    loss += cfg_train.weight_tv_density * model.density_total_variation(sdf_tv=0, smooth_grad_tv=smooth_grad_tv)
                if getattr(cfg_train, 'ori_tv', False):
                    loss += cfg_train.weight_tv_density * model.density_total_variation(sdf_tv=sdf_tv, smooth_grad_tv=0)
                    weight_tv_k0 = getattr(cfg_train, 'weight_tv_k0')
                    if weight_tv_k0 > 0:
                        k0_tv_terms = getattr(cfg_train, 'k0_tv_terms', dict())
                        loss += cfg_train.weight_tv_k0 * model.k0_total_variation(**k0_tv_terms)
                    if getattr(tv_terms, 'bg_density_tv', 0):
                        loss += cfg_train.weight_tv_density * model.density_total_variation(sdf_tv=0, smooth_grad_tv=0, bg_density_tv=tv_terms['bg_density_tv'])
        if getattr(cfg_train, 'ori_tv', False) and cfg_train.get('weight_bg_tv_k0', 0) >0 and global_step>cfg_train.tv_from and global_step%cfg_train.tv_every==0 and global_step<cfg_train.tv_end:
            bg_k0_tv_terms = getattr(cfg_train, 'bg_k0_tv_terms', dict())
            loss += cfg_train.get('weight_bg_tv_k0', 0) * model.bg_k0_total_variation(**bg_k0_tv_terms)

        if getattr(cfg_train, 'weight_rgb0', 0.) > 0:
            loss += F.mse_loss(render_result['rgb_marched0'], target) * cfg_train.weight_rgb0

        loss.backward()

        # make sure that density has no grad
        if global_step>cfg_train.tv_from and global_step<cfg_train.tv_end and global_step%cfg_train.tv_every==0:
            if not getattr(cfg_train, 'ori_tv', False):
                if cfg_train.weight_tv_density>0:
                    tv_terms = getattr(cfg_train, 'tv_terms', dict())
                    sdf_tv = tv_terms['sdf_tv']
                    if sdf_tv > 0:
                        model.sdf_total_variation_add_grad(
                            cfg_train.weight_tv_density * sdf_tv / len(rays_o), global_step < cfg_train.tv_dense_before)
                    bg_density_tv = getattr(tv_terms, 'bg_density_tv', 0)
                    if bg_density_tv > 0:
                        model.bg_density_total_variation_add_grad(
                            cfg_train.weight_tv_density * bg_density_tv / len(rays_o), global_step < cfg_train.tv_dense_before)                                                                                   
                if cfg_train.weight_tv_k0 > 0:
                    model.k0_total_variation_add_grad(
                        cfg_train.weight_tv_k0 / len(rays_o), global_step < cfg_train.tv_dense_before)
                if getattr(cfg_train, 'weight_bg_tv_k0', 0) > 0:
                    model.bg_k0_total_variation_add_grad(
                        cfg_train.weight_bg_tv_k0 / len(rays_o), global_step < cfg_train.tv_dense_before)
        optimizer.step()
```
![image](https://github.com/user-attachments/assets/4766ccf7-68fb-44f1-b2d3-bcadfdd55d77)

I have check the implementation of DVGOv2, which simply use CUDA after backward.
```
        loss.backward()

        if global_step<cfg_train.tv_before and global_step>cfg_train.tv_after and global_step%cfg_train.tv_every==0:
            if cfg_train.weight_tv_density>0:
                model.density_total_variation_add_grad(
                    cfg_train.weight_tv_density/len(rays_o), global_step<cfg_train.tv_dense_before)
            if cfg_train.weight_tv_k0>0:
                model.k0_total_variation_add_grad(
                    cfg_train.weight_tv_k0/len(rays_o), global_step<cfg_train.tv_dense_before)
```

Is it on purpose or simply a bug? But the performance is compatible with the paper. 
Thank you in advance!
